relay-event-schema/src/protocol/nel.rs

```diff
@@ -0,0 +1,302 @@
+//! Contains definitions for the Network Error Logging (NEL) interface.
+//!
+//! NEL is a browser feature that allows reporting of failed network requests from the client side.
+//! W3C Editor's Draft: <https://w3c.github.io/network-error-logging/>
+//! MDN: <https://developer.mozilla.org/en-US/docs/Web/HTTP/Network_Error_Logging>
+use std::ops::Sub;
+
+use chrono::{DateTime, Duration, Utc};
+use serde::{Deserialize, Serialize};
+
+use crate::processor::ProcessValue;
+use crate::protocol::{
+    AsPair, Event, HeaderName, HeaderValue, Headers, LogEntry, PairList, Request, TagEntry, Tags,
+    Timestamp,
+};
+#[cfg(feature = "jsonschema")]
+use relay_jsonschema_derive::JsonSchema;
+use relay_protocol::{Annotated, Empty, FromValue, IntoValue};
+use thiserror::Error;
+
+/// The NEL parsing errors.
+#[derive(Debug, Error)]
+pub enum NelError {
+    /// Unexpected format
+    #[error("unexpected format")]
+    InvalidNel,
+    /// Incoming Json is unparsable.
+    #[error("incoming json is unparsable")]
+    InvalidJson(#[from] serde_json::Error),
+}
+
+/// Inner (useful) part of a NEL report.
+///
+/// See `Nel` for meaning of fields.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+struct NelBodyRaw {
+    elapsed_time: Option<u64>,
+    method: Option<String>,
+    phase: Option<String>,
+    protocol: Option<String>,
+    referrer: Option<String>,
+    sampling_fraction: Option<f64>,
+    server_ip: Option<String>,
+    status_code: Option<u64>,
+    #[serde(rename = "type")]
+    ty: Option<String>,
+}
+
+/// Defines external, RFC-defined schema we accept, while `Nel` defines our own schema.
+///
+/// See `Nel` for meaning of fields.
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+struct NelReportRaw {
+    age: Option<i64>,
+    #[serde(rename = "type")]
+    ty: Option<String>, // always "network-error"
+    url: Option<String>,
+    user_agent: Option<String>,
+    body: NelBodyRaw,
+}
+
+impl NelReportRaw {
+    fn into_protocol(self) -> Nel {
```

Comment:
I see we do the same for security reports, but why are we deserializing into `NewlReportRaw` and then manually transforming into `Nel`? Is there no way to parse `Nel` directly?

Reply:
Removed `Raw` types and use `Nel` directly.